File: ios/Classes/Bridges/StoreTransactionBridge.swift
import Flutter
import SuperwallKit

extension StoreTransaction {
  func toJson() -> [String: Any] {
    var dict = [String: Any]()
    dict["configRequestId"] = configRequestId
    dict["appSessionId"] = appSessionId
    dict["triggerSessionId"] = triggerSessionId
    dict["transactionDate"] = transactionDate?.timeIntervalSince1970
    dict["originalTransactionIdentifier"] = originalTransactionIdentifier
    dict["state"] = state.rawValue
    dict["storeTransactionId"] = storeTransactionId
    dict["payment"] = payment.toJson()
    dict["originalTransactionDate"] = originalTransactionDate?.timeIntervalSince1970
    dict["webOrderLineItemID"] = webOrderLineItemID
    dict["appBundleId"] = appBundleId
    dict["subscriptionGroupId"] = subscriptionGroupId
    dict["isUpgraded"] = isUpgraded
    dict["expirationDate"] = expirationDate?.timeIntervalSince1970
    dict["offerId"] = offerId
    dict["revocationDate"] = revocationDate?.timeIntervalSince1970
    dict["appAccountToken"] = appAccountToken?.uuidString
    return dict
  }
}

extension StoreTransactionState {
  func toJson() -> String {
    switch self {
      case .purchasing:
        return "purchasing"
      case .purchased:
        return "purchased"
      case .failed:
        return "failed"
      case .restored:
        return "restored"
      case .deferred:
        return "deferred"
    }
  }
}

extension StorePayment {
  func toJson() -> [String: Any] {
    var dict = [String: Any]()
    dict["productIdentifier"] = productIdentifier
    dict["quantity"] = quantity
    if let discountIdentifier = discountIdentifier {
      dict["discountIdentifier"] = discountIdentifier
    }
    return dict
  }
}

File: android/src/main/kotlin/com/superwall/superwallkit_flutter/bridges/StoreTransactionBridge.kt
package com.superwall.superwallkit_flutter.bridges

import com.superwall.sdk.store.abstractions.transactions.StorePayment
import com.superwall.sdk.store.abstractions.transactions.StoreTransaction
import com.superwall.sdk.store.abstractions.transactions.StoreTransactionState

fun StoreTransaction.toJson(): Map<String, Any?> {
    return mapOf(
        "configRequestId" to configRequestId,
        "appSessionId" to appSessionId,
        "triggerSessionId" to triggerSessionId,
        "transactionDate" to transactionDate?.time,
        "originalTransactionIdentifier" to originalTransactionIdentifier,
        "state" to state.toJson(),
        "storeTransactionId" to storeTransactionId,
        "payment" to payment?.toJson(),
        "originalTransactionDate" to originalTransactionDate?.time,
        "webOrderLineItemID" to webOrderLineItemID,
        "appBundleId" to appBundleId,
        "subscriptionGroupId" to subscriptionGroupId,
        "isUpgraded" to isUpgraded,
        "expirationDate" to expirationDate?.time,
        "offerId" to offerId,
        "revocationDate" to revocationDate?.time,
        "appAccountToken" to appAccountToken?.toString()
    )
}

fun StoreTransactionState.toJson(): String {
    return when (this) {
        StoreTransactionState.Purchasing -> "purchasing"
        StoreTransactionState.Purchased -> "purchased"
        StoreTransactionState.Failed -> "failed"
        StoreTransactionState.Restored -> "restored"
        StoreTransactionState.Deferred -> "deferred"
    }
}

fun StorePayment.toJson(): Map<String, Any?> {
    return mapOf(
        // TODO Uncomment when implemented in Android
        // "productIdentifier" to productIdentifier,
        "quantity" to quantity,
        "discountIdentifier" to discountIdentifier
    )
}